webproject/src/main/java/com/google/sps/dataManagers/OrderManager.java

```diff
@@ -0,0 +1,107 @@
+
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.Query;
+import com.google.sps.data.OrderProto.Order;
+import com.google.sps.util.FirebaseUtil;
+
+@Singleton
+public class OrderManager {
+    private FirebaseUtil firebaseUtil;
+
+    @Inject
+    public OrderManager(FirebaseUtil firebaseUtil) {
+        this.firebaseUtil = firebaseUtil;
+    }
+
+    public Order addOrder(String userEmail, String product, int quantity, double unitPrice,
+                        String roomId) throws ServletException {
+        if (!isRoomIdValid(roomId)) {
+            throw new ServletException("Invalid roomId");
+        }
+     
+        String userEmailRoomId = userEmail + "_" + roomId;
+        double orderPrice = unitPrice * quantity;
+        
+        String quantityString = String.valueOf(quantity);
+        String unitPriceString = String.valueOf(unitPrice);
+        String orderPriceString = String.valueOf(orderPrice);
+        Map<String, String> orderMapping = new HashMap<>();
+        orderMapping.put("userEmail", userEmail);
+        orderMapping.put("product", product);        
+        orderMapping.put("quantity", quantityString);
+        orderMapping.put("unitPrice", unitPriceString);
+        orderMapping.put("orderPrice", orderPriceString);
+        orderMapping.put("roomId", roomId);
+        orderMapping.put("userEmailRoomId", userEmailRoomId);
+
+        firebaseUtil.getOrdersReference().push()
+                .setValue(orderMapping, (databaseError, databaseReference) -> {
+                    if (databaseError != null) {
+                        try {
+                            throw new ServletException("Invalid roomId");
+                        } catch (ServletException e) {
+                            e.printStackTrace();
+                        }
+                    }
+                });
+                
+        return Order.newBuilder()
+                    .setUserEmail(userEmail)
+                    .setProduct(product)
+                    .setQuantity(quantity)
+                    .setUnitPrice(unitPrice)
+                    .setOrderPrice(orderPrice)
+                    .setRoomId(roomId)
+                    .setUserEmailRoomId(userEmailRoomId)
+                    .build();
+    }
+
+    public List<Order> getMyOrders(String roomId, String userEmail) throws ServletException {
+        String userEmailRoomId = userEmail + "_" + roomId;
+        Query query = firebaseUtil.getOrdersReference().orderByChild("userEmailRoomId");
+        Optional<List<DataSnapshot>> dataSnapshots = firebaseUtil.getAllQuerySnapshots(query, userEmailRoomId);
+
+        List<Order> orders = new ArrayList<>();
+
+        if (!dataSnapshots.isPresent()) {
+            return null;
+        }
+
+        for (DataSnapshot dataSnapshot : dataSnapshots.get()) {
+            String product = dataSnapshot.child("product").getValue(String.class);
```

Comment:
move this to Order class, create a function called loadFromDataSnapshot.

Imagine order class get a new attribute, how does the person know that they need to change this line as well?

Reply:
I'm currently using an OrderProto which I'm not sure I can edit. Would it be ok to make that function in the OrderManager class instead?